modules/porous_flow/src/fvkernels/FVPorousFlowHeatAdvection.C

```diff
@@ -0,0 +1,103 @@
+  : FVFluxKernel(params),
+    _dictator(getUserObject<PorousFlowDictator>("PorousFlowDictator")),
+    _num_phases(_dictator.numPhases()),
+    _density(getADMaterialProperty<std::vector<Real>>("PorousFlow_fluid_phase_density_qp")),
+    _density_neighbor(
+        getNeighborADMaterialProperty<std::vector<Real>>("PorousFlow_fluid_phase_density_qp")),
+    _viscosity(getADMaterialProperty<std::vector<Real>>("PorousFlow_viscosity_qp")),
+    _viscosity_neighbor(
+        getNeighborADMaterialProperty<std::vector<Real>>("PorousFlow_viscosity_qp")),
+    _enthalpy(getADMaterialProperty<std::vector<Real>>("PorousFlow_fluid_phase_enthalpy_qp")),
+    _enthalpy_neighbor(
+        getNeighborADMaterialProperty<std::vector<Real>>("PorousFlow_fluid_phase_enthalpy_qp")),
+    _relperm(getADMaterialProperty<std::vector<Real>>("PorousFlow_relative_permeability_qp")),
+    _relperm_neighbor(
+        getNeighborADMaterialProperty<std::vector<Real>>("PorousFlow_relative_permeability_qp")),
+    _permeability(getADMaterialProperty<RealTensorValue>("PorousFlow_permeability_qp")),
+    _permeability_neighbor(
+        getNeighborADMaterialProperty<RealTensorValue>("PorousFlow_permeability_qp")),
+    _pressure(getADMaterialProperty<std::vector<Real>>("PorousFlow_porepressure_qp")),
+    _pressure_neighbor(
+        getNeighborADMaterialProperty<std::vector<Real>>("PorousFlow_porepressure_qp")),
+    _grad_p(getADMaterialProperty<std::vector<RealGradient>>("PorousFlow_grad_porepressure_qp")),
+    _gravity(getParam<RealVectorValue>("gravity"))
+{
+}
+
+ADReal
+FVPorousFlowHeatAdvection::computeQpResidual()
+{
+  ADReal flux = 0.0;
+  ADRealGradient pressure_grad;
+  ADRealTensorValue mobility;
+
+  for (unsigned int p = 0; p < _num_phases; ++p)
+  {
+    // If we are on a boundary face, use the reconstructed gradient computed in _grad_p
+    if (onBoundary(*_face_info))
+    {
+      const auto & gradp = -_grad_p[_qp][p];
+      pressure_grad = gradp + _density[_qp][p] * _gravity;
+
+      mobility = _enthalpy[_qp][p] * _relperm[_qp][p] * _permeability[_qp] * _density[_qp][p] /
+                 _viscosity[_qp][p];
+    }
+    else
+    {
+      // If we are on an internal face, calculate the gradient explicitly
+      const auto & p_elem = _pressure[_qp][p];
+      const auto & p_neighbor = _pressure_neighbor[_qp][p];
+
+      const auto gradp = (p_elem - p_neighbor) * _face_info->eCN() / _face_info->dCNMag();
+
+      const auto mobility_element = _enthalpy[_qp][p] * _relperm[_qp][p] * _permeability[_qp] *
+                                    _density[_qp][p] / _viscosity[_qp][p];
+
+      const auto mobility_neighbor = _enthalpy_neighbor[_qp][p] * _relperm_neighbor[_qp][p] *
+                                     _permeability_neighbor[_qp] * _density_neighbor[_qp][p] /
+                                     _viscosity_neighbor[_qp][p];
+
+      pressure_grad = gradp + _density[_qp][p] * _gravity;
```

Comment:
any reason the density term is centered on the element instead of the face?

Reply:
Only that I don't know how to specify the face density 😄

Reply:
should be the same as you did for the others? some weighting of density and density_neihgbor

Reply:
Oh, I get what you mean now! I thought that you meant there was something like `_density[_face]` rather than `_density[_qp]`